File: web/src/lib/audit.ts
/**
 * The audit log (US-B1-02, US-B9-04, decision 53r). Every administrator or
 * evaluator action that changes anything goes through here, with a reason.
 * "Unlimited authority without a record is what disputes feed on."
 */
import { db } from "@/db";
import { auditLog } from "@/db/schema";

type Tx = Parameters<Parameters<typeof db.transaction>[0]>[0];

export async function audit(
  entry: {
    actorId: string | null;
    action: string;
    target?: string | null;
    reason: string;
    detail?: unknown;
  },
  tx: Tx | typeof db = db,
): Promise<void> {
  await tx.insert(auditLog).values({
    actorId: entry.actorId,
    action: entry.action,
    target: entry.target ?? null,
    reason: entry.reason,
    detail: entry.detail ?? null,
  });
}

File: engine/engine/core/audit.py
"""The audit log. Every staff action that changes anything writes a row here,
with a reason, inside the same transaction as the change -- so there is no state
in which the contest moved and the log does not say who moved it.
"""

from __future__ import annotations

from typing import Any

import sqlalchemy as sa

from engine.schema import audit_log


def audit(
    conn: sa.Connection,
    *,
    actor_id: str | None,
    action: str,
    reason: str,
    target: str | None = None,
    detail: Any = None,
) -> None:
    conn.execute(
        sa.insert(audit_log).values(
            actor_id=actor_id, action=action, target=target, reason=reason, detail=detail
        )
    )
